### PR TITLE
VACMS-9249: Add publish function for update hooks.

### LINKED ISSUE #9249: Add node publisher function to va_gov_db install.
## User Story or Problem Statement

As a developer I need a function that will publish a set of specified nodes so that knowledgebase articles or other nodes can be published as part of a code release.

## Acceptance Criteria
- [x] function lives in va_gov_db.install
   - [x] accepts an array of nids to publish
   - [x] adds the revision log to say "Automatically published by code update."
   - [x] creates a new revision from the most recent and publishes it
   - [x] throws update exception if node not found.
- [ ] returns a message of confirmation of node titles pubilished to be returned by hoook_update_n

## Implementation steps
<optional>

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
closes #9249

## Testing done
Visual / drush updb

## Screenshots
image

## QA steps
 - [ ] Checkout this pr branch locally
 - [ ] Go to `/node/6457/how-to-edit-an-event`
 - [ ] Edit the latest revision title and save as draft
 - [ ] Add the code snippet below to va_gov_db.install
 ```
/**
 * Publish KB events article.
 */
function va_gov_db_update_9006(&$sandbox) {
  $nids = [6483];
  return _va_gov_db_publish_nodes($nids);
}
```
- [ ] Run `drush updb`
- [ ] Verify terminal message akin to one in screenshot
- [ ] Go to `/node/6457/how-to-edit-an-event` and visually verify your changes saved, and have been published by `CMS migrator` as the latest revision

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: docroot/modules/custom/va_gov_db/va_gov_db.install [hunk 1]
```diff
@@ -224,6 +224,52 @@ function _va_gov_db_create_redirect_by_path($old_path, $new_path) {
...
+    $latest_vid_loaded->status = 1;
+    $latest_vid_loaded->revision_log = t("Automatically published by code update.");
+    $latest_vid_loaded->set('moderation_state', "published");
+    $latest_vid_loaded->setNewRevision();
+    $latest_vid_loaded->isDefaultRevision(TRUE);
+    $latest_vid_loaded->setRevisionUserId(1317);
+    $latest_vid_loaded->setOwnerId(1317);
+    $latest_vid_loaded->setRevisionCreationTime(time());
+    $latest_vid_loaded->save();
+    $nids_published[] = $nid;
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_db/va_gov_db.install:266]:
```suggestion
    $nodes_published[] = "{$nid} - {$latest_vid_loaded->getTitle()}";
```

**Review by swirtSJW — CHANGES_REQUESTED** [review]
Good start.  Needs some de-fanging and slightly missed an AC.

File: docroot/modules/custom/va_gov_db/va_gov_db.install [hunk 2]
```diff
@@ -224,6 +224,52 @@ function _va_gov_db_create_redirect_by_path($old_path, $new_path) {
...
+ * @throws Drupal\Core\Utility\UpdateException
+ */
+function _va_gov_db_publish_nodes(array $nids) {
+  $nids_published = [];
+  foreach ($nids as $nid) {
+    // Get latest revision ID.
+    $node_storage = \Drupal::entityTypeManager()
+      ->getStorage('node');
+    $latest_vid = $node_storage->getLatestRevisionId($nid);
+    $blah = '';
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_db/va_gov_db.install]: Lol - before I started reading your review, I looked at my editor and saw this. Thought I was gonna rebase it in before you were any the wiser, but you busted me.

File: docroot/modules/custom/va_gov_db/va_gov_db.install (hunk 1 above)
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_db/va_gov_db.install:263]: D'oh. My bad, @swirtSJW

**Review by swirtSJW — APPROVED** [review]
Nice work on this.

**Pull request merged**